pandas/io/formats/excel.py

```diff
@@ -635,35 +638,45 @@ def _format_hierarchical_rows(self):
                         if spans[i] > 1:
                             yield ExcelCell(
                                 self.rowcounter + i,
-                                gcolidx,
+                                colidx,
                                 values[i],
                                 self.header_style,
                                 self.rowcounter + i + spans[i] - 1,
-                                gcolidx,
+                                colidx,
                             )
                         else:
                             yield ExcelCell(
                                 self.rowcounter + i,
-                                gcolidx,
+                                colidx,
                                 values[i],
                                 self.header_style,
                             )
-                    gcolidx += 1
+                    colidx += 1
+
+                for cell in self._generate_body(colidx):
+                    yield cell
 
             else:
-                # Format hierarchical rows with non-merged values.
-                for indexcolvals in zip(*self.df.index):
-                    for idx, indexcolval in enumerate(indexcolvals):
+                _, ncol = self.df.shape
```

Comment:
what are all of these changes about?

Reply:
The previous code will generate `ExcelCell` of `df.index` column by column and then `ExcelCell` of `df.values` column by column. Now generate the first row of `ExcelCell` of `df.index` and `df.values`, then the second row and so on.

PS: `gcolidx` is revised to `colidx` for consistency only, as I use `rowidx` to represent row index.

Reply:
The approach is consistent with what was suggested in the original issue:

https://github.com/pandas-dev/pandas/issues/15392#issuecomment-279699881